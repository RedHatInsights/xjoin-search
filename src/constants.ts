export const IDENTITY_HEADER = 'x-rh-identity';

export const ES_NULL_VALUE = '$$_XJOIN_SEARCH_NULL_VALUE';
