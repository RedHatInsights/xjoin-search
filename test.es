

GET /test.hosts.v1/_mapping

GET /test.hosts/_doc/_search
