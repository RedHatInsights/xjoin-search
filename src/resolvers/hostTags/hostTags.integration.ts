import { runQuery, createHeaders } from '../../../test';
import * as constants from '../../constants';
import createIdentityHeader from '../../middleware/identity/utils';

const TAG_FILTERS_QUERY = `
    query hostTags (
        $hostFilter: HostFilter,
        $filter: TagAggregationFilter,
        $order_by: HOST_TAGS_ORDER_BY,
        $order_how: ORDER_DIR,
        $limit: Int,
        $offset: Int) {
        hostTags (
            hostFilter: $hostFilter,
            filter: $filter,
            order_by: $order_by,
            order_how: $order_how,
            limit: $limit,
            offset: $offset
        )
        {
            meta {
                count,
                total
            }
            data {
                tag {
                    namespace, key, value
                },
                count
            }
        }
    }
`;

describe('host tags', function () {
    test('basic query', async () => {
        const { data, status } = await runQuery(TAG_FILTERS_QUERY, {});
        expect(status).toEqual(200);
        expect(data).toMatchSnapshot();
    });

    test('account isolation', async () => {
        const headers = {
            [constants.IDENTITY_HEADER]: createIdentityHeader(f => f, 'customer', '12345', false)
        };

        const { data, status } = await runQuery(TAG_FILTERS_QUERY, {}, headers);
        data.hostTags.data.should.have.length(0);
        data.hostTags.meta.should.have.property('total', 0);
        data.hostTags.meta.should.have.property('count', 0);
        expect(status).toEqual(200);
    });

    describe('pagination', function () {
        const headers = createHeaders('hostTagsTest', 'hostTagsTest');

        test('limit', async () => {
            const { data, status } = await runQuery(TAG_FILTERS_QUERY, {
                limit: 2
            }, headers);

            expect(status).toEqual(200);
            data.hostTags.data.should.have.length(2);
            data.hostTags.meta.count.should.equal(2);
            data.hostTags.meta.total.should.equal(16);
            expect(data).toMatchSnapshot();
        });

        test('limit + offset', async () => {
            const { data, status } = await runQuery(TAG_FILTERS_QUERY, {
                limit: 2,
                offset: 14
            }, headers);

            data.hostTags.data.should.have.length(2);
            data.hostTags.meta.should.have.property('count', 2);
            data.hostTags.meta.should.have.property('total', 16);

            expect(status).toEqual(200);
            expect(data).toMatchSnapshot();
        });
    });

    test('ordering', async () => {
        const { data, status } = await runQuery(TAG_FILTERS_QUERY, {
            order_by: 'tag',
            order_how: 'ASC'
        });

        expect(status).toEqual(200);
        expect(data).toMatchSnapshot();
    });

    describe('host filters', function () {
        test('by id', async () => {
            const { data, status } = await runQuery(TAG_FILTERS_QUERY, {
                hostFilter: { id: { eq: '22cd8e39-13bb-4d02-8316-84b850dc5136' }}
            });

            expect(status).toEqual(200);
            expect(data).toMatchSnapshot();
        });

        test('by tag', async () => {
            const { data, status } = await runQuery(TAG_FILTERS_QUERY, {
                hostFilter: {
                    tag: {
                        namespace: {eq: 'insights-client'},
                        key: {eq: 'database'}
                    }
                }
            });

            expect(status).toEqual(200);
            expect(data).toMatchSnapshot();
        });

        test('by tag intersection', async () => {
            const { data, status } = await runQuery(TAG_FILTERS_QUERY, {
                hostFilter: {
                    AND: [{
                        tag: {
                            namespace: {eq: 'insights-client'},
                            key: {eq: 'os'},
                            value: {eq: 'fedora'}
                        }
                    }, {
                        tag: {
                            namespace: {eq: 'insights-client'},
                            key: {eq: 'database'}
                        }
                    }]
                }
            });

            expect(status).toEqual(200);
            expect(data).toMatchSnapshot();
        });
    });

    describe('tag filters', function () {
        test('by tag name', async () => {
            const { data, status } = await runQuery(TAG_FILTERS_QUERY, {
                filter: {
                    search: {
                        eq: 'aws/region=us-east-1'
                    }
                }
            });

            expect(status).toEqual(200);
            data.hostTags.data.should.have.length(1);
            data.hostTags.meta.should.have.property('count', 1);
            data.hostTags.meta.should.have.property('total', 8);
            data.hostTags.data.should.eql([{
                count: 1,
                tag: {
                    namespace: 'aws',
                    key: 'region',
                    value: 'us-east-1'
                }
            }]);
        });

        test('by tag name negative', async () => {
            const { data, status } = await runQuery(TAG_FILTERS_QUERY, {
                filter: {
                    search: {
                        eq: '.*aws.*'
                    }
                }
            });

            expect(status).toEqual(200);
            data.hostTags.data.should.be.empty();
        });

        test('by tag name prefix', async () => {
            const { data, status } = await runQuery(TAG_FILTERS_QUERY, {
                filter: {
                    search: {
                        regex: 'insights-client.*'
                    }
                }
            });

            expect(status).toEqual(200);
            expect(data).toMatchSnapshot();
        });

        test('by tag name substring', async () => {
            const { data, status } = await runQuery(TAG_FILTERS_QUERY, {
                filter: {
                    search: {
                        regex: '.*region.*'
                    }
                }
            });

            expect(status).toEqual(200);
            expect(data).toMatchSnapshot();
        });

        test('by tag name suffix', async () => {
            const { data, status } = await runQuery(TAG_FILTERS_QUERY, {
                filter: {
                    search: {
                        regex: '.*Δwithčhars!'
                    }
                }
            });

            expect(status).toEqual(200);
            expect(data).toMatchSnapshot();
        });

        describe('special characters', function () {
            const headers = {
                [constants.IDENTITY_HEADER]: createIdentityHeader(f => f, 'hostTagsSpecialChars', 'hostTagsSpecialChars', false)
            };

            const specialTags = [{
                count: 1,
                tag: {
                    namespace: 'insights-client',
                    key: 'key',
                    value: 'keyΔwithčhars*+!.,-_ '
                }
            }, {
                count: 1,
                tag: {
                    namespace: 'insights-client',
                    key: 'keyΔwithčhars*+!.,-_ ',
                    value: 'value'
                }
            }, {
                count: 1,
                tag: {
                    namespace: 'keyΔwithčhars*+!.,-_ ',
                    key: 'key',
                    value: 'value'
                }
            }];

            Array.from('Δč*+!.,_ ').forEach(i =>
                test(`search by "${i}"`, async () => {
                    const { data, status } = await runQuery(TAG_FILTERS_QUERY, {
                        filter: {
                            search: {
                                regex: `.*\\${i}.*`
                            }
                        }
                    }, headers);

                    expect(status).toEqual(200);
                    const d = data.hostTags.data;
                    d.should.have.length(3);
                    d.should.eql(specialTags);
                })
            );

            Array.from('=/').forEach(i =>
                test(`search by control characters ("${i}") returns all tags`, async () => {
                    const { data, status } = await runQuery(TAG_FILTERS_QUERY, {
                        filter: {
                            search: {
                                regex: `.*\\${i}.*`
                            }
                        }
                    }, headers);

                    expect(status).toEqual(200);
                    const d = data.hostTags.data;
                    d.should.have.length(4);
                })
            );
        });
    });
});
