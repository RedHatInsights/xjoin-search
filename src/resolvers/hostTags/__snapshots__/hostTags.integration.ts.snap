// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`host tags basic query 1`] = `
Object {
  "hostTags": Object {
    "data": Array [
      Object {
        "count": 3,
        "tag": Object {
          "key": "env",
          "namespace": "Sat",
          "value": "prod",
        },
      },
      Object {
        "count": 2,
        "tag": Object {
          "key": "database",
          "namespace": "insights-client",
          "value": null,
        },
      },
      Object {
        "count": 2,
        "tag": Object {
          "key": "os",
          "namespace": "insights-client",
          "value": "fedora",
        },
      },
      Object {
        "count": 1,
        "tag": Object {
          "key": "region",
          "namespace": "aws",
          "value": "us-east-1",
        },
      },
      Object {
        "count": 1,
        "tag": Object {
          "key": "region",
          "namespace": "aws",
          "value": "us-east-2",
        },
      },
      Object {
        "count": 1,
        "tag": Object {
          "key": "region",
          "namespace": "aws",
          "value": "us-east-3",
        },
      },
      Object {
        "count": 1,
        "tag": Object {
          "key": "special/keyΔwithčhars",
          "namespace": "insights-client",
          "value": "special/valueΔwithčhars!",
        },
      },
      Object {
        "count": 1,
        "tag": Object {
          "key": "web",
          "namespace": "insights-client",
          "value": null,
        },
      },
    ],
    "meta": Object {
      "count": 8,
      "total": 8,
    },
  },
}
`;

exports[`host tags host filters by id 1`] = `
Object {
  "hostTags": Object {
    "data": Array [
      Object {
        "count": 1,
        "tag": Object {
          "key": "env",
          "namespace": "Sat",
          "value": "prod",
        },
      },
      Object {
        "count": 1,
        "tag": Object {
          "key": "region",
          "namespace": "aws",
          "value": "us-east-2",
        },
      },
      Object {
        "count": 1,
        "tag": Object {
          "key": "os",
          "namespace": "insights-client",
          "value": "fedora",
        },
      },
      Object {
        "count": 1,
        "tag": Object {
          "key": "special/keyΔwithčhars",
          "namespace": "insights-client",
          "value": "special/valueΔwithčhars!",
        },
      },
      Object {
        "count": 1,
        "tag": Object {
          "key": "web",
          "namespace": "insights-client",
          "value": null,
        },
      },
    ],
    "meta": Object {
      "count": 5,
      "total": 5,
    },
  },
}
`;

exports[`host tags host filters by tag 1`] = `
Object {
  "hostTags": Object {
    "data": Array [
      Object {
        "count": 2,
        "tag": Object {
          "key": "env",
          "namespace": "Sat",
          "value": "prod",
        },
      },
      Object {
        "count": 2,
        "tag": Object {
          "key": "database",
          "namespace": "insights-client",
          "value": null,
        },
      },
      Object {
        "count": 1,
        "tag": Object {
          "key": "region",
          "namespace": "aws",
          "value": "us-east-1",
        },
      },
      Object {
        "count": 1,
        "tag": Object {
          "key": "region",
          "namespace": "aws",
          "value": "us-east-3",
        },
      },
      Object {
        "count": 1,
        "tag": Object {
          "key": "os",
          "namespace": "insights-client",
          "value": "fedora",
        },
      },
    ],
    "meta": Object {
      "count": 5,
      "total": 5,
    },
  },
}
`;

exports[`host tags host filters by tag intersection 1`] = `
Object {
  "hostTags": Object {
    "data": Array [
      Object {
        "count": 1,
        "tag": Object {
          "key": "env",
          "namespace": "Sat",
          "value": "prod",
        },
      },
      Object {
        "count": 1,
        "tag": Object {
          "key": "region",
          "namespace": "aws",
          "value": "us-east-3",
        },
      },
      Object {
        "count": 1,
        "tag": Object {
          "key": "database",
          "namespace": "insights-client",
          "value": null,
        },
      },
      Object {
        "count": 1,
        "tag": Object {
          "key": "os",
          "namespace": "insights-client",
          "value": "fedora",
        },
      },
    ],
    "meta": Object {
      "count": 4,
      "total": 4,
    },
  },
}
`;

exports[`host tags ordering 1`] = `
Object {
  "hostTags": Object {
    "data": Array [
      Object {
        "count": 3,
        "tag": Object {
          "key": "env",
          "namespace": "Sat",
          "value": "prod",
        },
      },
      Object {
        "count": 1,
        "tag": Object {
          "key": "region",
          "namespace": "aws",
          "value": "us-east-1",
        },
      },
      Object {
        "count": 1,
        "tag": Object {
          "key": "region",
          "namespace": "aws",
          "value": "us-east-2",
        },
      },
      Object {
        "count": 1,
        "tag": Object {
          "key": "region",
          "namespace": "aws",
          "value": "us-east-3",
        },
      },
      Object {
        "count": 2,
        "tag": Object {
          "key": "database",
          "namespace": "insights-client",
          "value": null,
        },
      },
      Object {
        "count": 2,
        "tag": Object {
          "key": "os",
          "namespace": "insights-client",
          "value": "fedora",
        },
      },
      Object {
        "count": 1,
        "tag": Object {
          "key": "special/keyΔwithčhars",
          "namespace": "insights-client",
          "value": "special/valueΔwithčhars!",
        },
      },
      Object {
        "count": 1,
        "tag": Object {
          "key": "web",
          "namespace": "insights-client",
          "value": null,
        },
      },
    ],
    "meta": Object {
      "count": 8,
      "total": 8,
    },
  },
}
`;

exports[`host tags pagination limit + offset 1`] = `
Object {
  "hostTags": Object {
    "data": Array [
      Object {
        "count": 1,
        "tag": Object {
          "key": "foo",
          "namespace": "insights-client",
          "value": "7",
        },
      },
      Object {
        "count": 1,
        "tag": Object {
          "key": "foo",
          "namespace": "insights-client",
          "value": "8",
        },
      },
    ],
    "meta": Object {
      "count": 2,
      "total": 16,
    },
  },
}
`;

exports[`host tags pagination limit 1`] = `
Object {
  "hostTags": Object {
    "data": Array [
      Object {
        "count": 1,
        "tag": Object {
          "key": "bar",
          "namespace": "insights-client",
          "value": "1",
        },
      },
      Object {
        "count": 1,
        "tag": Object {
          "key": "bar",
          "namespace": "insights-client",
          "value": "2",
        },
      },
    ],
    "meta": Object {
      "count": 2,
      "total": 16,
    },
  },
}
`;

exports[`host tags tag filters by tag name prefix 1`] = `
Object {
  "hostTags": Object {
    "data": Array [
      Object {
        "count": 2,
        "tag": Object {
          "key": "database",
          "namespace": "insights-client",
          "value": null,
        },
      },
      Object {
        "count": 2,
        "tag": Object {
          "key": "os",
          "namespace": "insights-client",
          "value": "fedora",
        },
      },
      Object {
        "count": 1,
        "tag": Object {
          "key": "special/keyΔwithčhars",
          "namespace": "insights-client",
          "value": "special/valueΔwithčhars!",
        },
      },
      Object {
        "count": 1,
        "tag": Object {
          "key": "web",
          "namespace": "insights-client",
          "value": null,
        },
      },
    ],
    "meta": Object {
      "count": 4,
      "total": 8,
    },
  },
}
`;

exports[`host tags tag filters by tag name substring 1`] = `
Object {
  "hostTags": Object {
    "data": Array [
      Object {
        "count": 1,
        "tag": Object {
          "key": "region",
          "namespace": "aws",
          "value": "us-east-1",
        },
      },
      Object {
        "count": 1,
        "tag": Object {
          "key": "region",
          "namespace": "aws",
          "value": "us-east-2",
        },
      },
      Object {
        "count": 1,
        "tag": Object {
          "key": "region",
          "namespace": "aws",
          "value": "us-east-3",
        },
      },
    ],
    "meta": Object {
      "count": 3,
      "total": 8,
    },
  },
}
`;

exports[`host tags tag filters by tag name suffix 1`] = `
Object {
  "hostTags": Object {
    "data": Array [
      Object {
        "count": 1,
        "tag": Object {
          "key": "special/keyΔwithčhars",
          "namespace": "insights-client",
          "value": "special/valueΔwithčhars!",
        },
      },
    ],
    "meta": Object {
      "count": 1,
      "total": 8,
    },
  },
}
`;
